src/data/equiv/derangements/basic.lean

```diff
@@ -0,0 +1,265 @@
+def derangement_congr {α β : Type*} (e : α ≃ β) : (derangements α ≃ derangements β) :=
+begin
+  refine subtype_equiv (perm_congr e) _,
+  intro f,
+  simp only [derangements, is_derangement, perm_congr_apply, set.mem_set_of_eq],
+  split,
+  { contrapose!,
+    rintro ⟨y, hy_fixed⟩,
+    use e.symm y,
+    apply e.injective,
+    simp [hy_fixed] },
+  { contrapose!,
+    rintro ⟨x, hx_fixed⟩,
+    use e x,
+    simp [hx_fixed] }
+end
+
+end TODO_name
+
+section fixed_points
+
+variables {α : Type*}
+
+lemma eofp_iff_opfp_and_eq (a : α) (f : perm α) :
+  exactly_one_fixed_point f a ↔ only_possible_fixed_point f a ∧ f a = a :=
+begin
+  split,
+  { intro h_eofp, exact ⟨λ x, (h_eofp x).mp, (h_eofp a).mpr rfl⟩ },
+  { rintros ⟨h_opfp, fa_eq_a⟩ x,
+    split,
+    { exact h_opfp x },
+    { rintro ⟨rfl⟩, exact fa_eq_a }}
+end
+
+lemma derangement_iff_opfp_and_ne (a : α) (f : perm α) :
+  is_derangement f ↔ only_possible_fixed_point f a ∧ f a ≠ a :=
+begin
+  unfold is_derangement only_possible_fixed_point,
+  split,
+  { intro h_derangement,
+    split,
+    { intros x x_fixed, exfalso, exact h_derangement x x_fixed },
+    { exact h_derangement a }},
+  { rintros ⟨h_opfp, fa_ne_a⟩ x x_fixed,
+    specialize h_opfp x x_fixed,
+    rw h_opfp at x_fixed,
+    exact fa_ne_a x_fixed }
+end
+
+/-- The set of permutations with `a` the only fixed point is equivalent to the set of derangements
+    on `{a}ᶜ`. -/
+def eofp_equiv_derangements_except_for [decidable_eq α] (a : α) :
+  {f : perm α | exactly_one_fixed_point f a} ≃ derangements ({a}ᶜ : set α) :=
+begin
+  let key_equiv : {f : perm α | f a = a} ≃ perm ({a}ᶜ : set α),
+  { transitivity {f : perm α // ∀ x : ({a} : set α), f x = (equiv.refl _) x},
+    { apply subtype_equiv_right, simp },
+    { exact equiv.set.compl (equiv.refl _) }},
+
+  transitivity {f : {f : perm α // f a = a} // only_possible_fixed_point f.val a},
+  { apply equiv.trans _ (subtype_subtype_equiv_subtype_exists _ _).symm,
+    apply subtype_equiv_right,
+    intro f,
+    simp only [exists_prop, set.mem_set_of_eq],
+    rw and_comm,
+    exact eofp_iff_opfp_and_eq _ _ },
+  { refine subtype_equiv key_equiv _,
+    rintro ⟨f, _⟩,
+    simp [key_equiv, equiv.set.compl, derangements, is_derangement,
+          only_possible_fixed_point, not_imp_not] }
+end
+
+end fixed_points
+
+section option
+
+variables {α : Type*}
+
+lemma remove_none_fixed_pt (f : perm (option α)) (x : α) :
+  (remove_none f) x = x ↔ f (some x) = some x ∨ (f none = some x ∧ f (some x) = none) :=
+begin
+  -- allows remove_none_XXX lemmas to work
+  rw ← option.some_inj,
+  cases fx : f (some x) with y,
+  { rw remove_none_none f fx, simp },
+  { rw [remove_none_some f ⟨y, fx⟩, fx], simp }
+end
+
+-- TODO I defined this predicate `foo` and the corresponding set, but it's only
+-- used in the proof of `derangements_equiv_sigma_opfp`. Should it be completely
+-- folded into that proof?
```

Comment:
`foo_lemma` is clearly much too long to inline.

Reply:
yeah, guess i'll cook up a name of some kind. probably should wrap it behind a namespace too